src/Control/Effect/Telemetry.hs

```diff
@@ -0,0 +1,49 @@
+module Control.Effect.Telemetry (
+  Telemetry (..),
+  trackUsage,
+  trackTimeSpent,
+  trackConfig,
+  trackRawLogMessage,
+  setSink,
+  trackResult,
+) where
+
+import Control.Algebra (Has, send)
+import Control.Carrier.Telemetry.Types (CountableCliFeature, TelemetrySink)
+import Data.Aeson (ToJSON (toJSON), Value)
+import Data.Text (Text)
+import Diag.Result (Result (Failure, Success), renderFailure, renderSuccess)
+import Effect.Logger (Severity (SevError, SevWarn), renderIt)
+import Prettyprinter (Doc, unAnnotate)
+import Prettyprinter.Render.Terminal (AnsiStyle)
+
+data Telemetry m a where
+  TrackUsage :: CountableCliFeature -> Telemetry m ()
+  TrackTimeSpent :: Text -> m a -> Telemetry m a
+  TrackConfig :: Text -> Value -> Telemetry m ()
+  TrackRawLogMessage :: Severity -> Text -> Telemetry m ()
+  SetTelemetrySink :: TelemetrySink -> Telemetry m ()
+
+trackUsage :: Has Telemetry sig m => CountableCliFeature -> m ()
+trackUsage = send . TrackUsage
+
+trackTimeSpent :: Has Telemetry sig m => Text -> m a -> m a
+trackTimeSpent header act = send $ TrackTimeSpent header act
```

Comment:
[nit]: reduce to point-free (hlint should have caught these, not sure why it didn't)
```suggestion
trackTimeSpent header = send . TrackTimeSpent header
```